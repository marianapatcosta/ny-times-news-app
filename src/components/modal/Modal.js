import { defineComponent } from "vue";

export default defineComponent({
  name: "Modal",
  emits: ["close"],
  props: {
    header: {
      type: String,
      required: true,
    },
    message: {
      type: String,
    },
  },
  methods: {
    handleClickOutside(event) {
      const target = event.target;
      const modal = this.$refs.modal;
      if (modal && !modal.contains(target)) {
        this.$emit("close");
      }
    },
  },
  mounted() {
    document.addEventListener("click", this.handleClickOutside);
  },
  beforeUnmount() {
    document.removeEventListener("click", this.handleClickOutside);
  },
});
