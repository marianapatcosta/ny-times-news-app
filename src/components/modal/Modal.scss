.modal {
  &__overlay {
    height: 100vh;
    width: 100vw;
    position: fixed;
    top: 0;
    left: 0;
    background-color: rgba($color-black, 0.7);
    display: flex;
    z-index: 100;
  }

  &__content {
    margin: auto;
    font-size: 90%;
    background-color: $color-primary;
    box-shadow: 0 0.125rem 0.5rem rgba($color-black, 0.7);
    min-width: 17rem;
  }

  &__header {
    padding: 0.625rem;
    background-color: $color-secondary;
    color: $color_primary;
    position: relative;
    display: flex;
    justify-content: space-between;

    button {
      cursor: pointer;
      transition: opacity 0.2s ease-in-out;
      -webkit-transition: opacity 0.2s ease-in-out;
      -moz-transition: opacity 0.2s ease-in-out;
      -ms-transition: opacity 0.2s ease-in-out;
      -o-transition: opacity 0.2s ease-in-out;

      &:hover {
        opacity: 0.75;
      }

      &:focus {
        outline: 0.125rem solid $color_primary;
      }

      &:focus:not(:focus-visible) {
        outline: none;
      }
    }
  }

  &__message {
    margin-top: 0.625rem;
    padding: 1.5rem 0.625rem;
    text-align: center;
  }

  &__footer {
    display: flex;
    align-items: center;
    justify-content: center;
    padding-bottom: 0.5rem;

    button {
      background-color: $color-secondary;
      color: $color_primary;
      padding: 0.3rem 1rem;
      border-radius: 0.25rem;
      -webkit-border-radius: 0.25rem;
      -moz-border-radius: 0.25rem;
      -ms-border-radius: 0.25rem;
      -o-border-radius: 0.25rem;
    }
  }
}

@media only screen and (min-width: $xs) {
  .modal {
    &__content {
      min-width: 20rem;
    }
  }
}
