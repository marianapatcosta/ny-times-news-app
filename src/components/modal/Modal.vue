<template>
  <div class="modal">
    <div class="modal__overlay">
      <div class="modal__content" ref="modal">
        <header class="modal__header">
          <h4>{{ header }}</h4>
          <button
            class="modal__close-button"
            @click.prevent="$emit('close')"
            @keydown.enter="$emit('close')"
          >
            x
          </button>
        </header>
        <main>
          <p class="modal__message">{{ message }}</p>
        </main>
        <footer class="modal__footer">
          <button
            class="modal__ok-button"
            aria-label="Click to close the error modal."
            @click.prevent="$emit('close')"
            @keydown.enter="$emit('close')"
          >
            ok
          </button>
        </footer>
      </div>
    </div>
  </div>
</template>

<script src="./Modal.js"></script>

<style scoped lang="scss" src="./Modal.scss"></style>
