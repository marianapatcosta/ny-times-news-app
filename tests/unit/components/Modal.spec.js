import { mount } from "@vue/test-utils";
import Modal from "@/components/modal/Modal.vue";

describe("Modal.vue", () => {
  const header = "Modal header";
  let wrapper;
  beforeEach(() => {
    wrapper = mount(Modal, {
      props: {
        header,
      },
    });
  });

  afterEach(() => {
    wrapper.unmount();
  });
  it("should render without errors and match snapshot renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should render props.header when passed", () => {
    expect(wrapper.text()).toContain(header);
  });

  it("should emit 'close' event when x is clicked", async () => {
    const closeButton = wrapper.find(".modal__close-button");
    await closeButton.trigger("click");
    expect(wrapper.emitted()).toHaveProperty("close");
  });

  it("should emit 'close' event when 'ok' is clicked", async () => {
    const closeButton = wrapper.find(".modal__ok-button");
    await closeButton.trigger("click");
    expect(wrapper.emitted()).toHaveProperty("close");
  });

  it("should not emit 'close' event when modal area is clicked", async () => {
    const div = document.createElement("div");
    document.body.appendChild(div);
    const handleClickOutsideFn = jest.spyOn(
      Modal.methods,
      "handleClickOutside"
    );
    await wrapper.trigger("click");
    expect(handleClickOutsideFn).not.toHaveBeenCalled();
  });

  it("should emit 'close' event when any area outside the modal is clicked", async () => {
    const handleClickOutsideFn = jest.spyOn(
      Modal.methods,
      "handleClickOutside"
    );
    await document.dispatchEvent(new Event("click"));
    expect(handleClickOutsideFn).toHaveBeenCalled();
  });

  it("should remove close event listener when beforeMounted is executed", async () => {
    Object.defineProperty(document, "removeEventListener", {
      value: jest.fn(),
    });
    await wrapper.unmount();
    expect(document.removeEventListener).toHaveBeenCalled();
    expect(document.removeEventListener).toHaveBeenCalledWith(
      "click",
      wrapper.vm.handleClickOutside
    );
  });
});
