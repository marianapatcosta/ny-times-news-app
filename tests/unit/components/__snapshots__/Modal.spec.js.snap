// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal.vue should render without errors and match snapshot renders correctly 1`] = `
<div
  class="modal"
>
  <div
    class="modal__overlay"
  >
    <div
      class="modal__content"
    >
      <header
        class="modal__header"
      >
        <h4>
          Modal header
        </h4>
        <button
          class="modal__close-button"
        >
           x 
        </button>
      </header>
      <main>
        <p
          class="modal__message"
        />
      </main>
      <footer
        class="modal__footer"
      >
        <button
          aria-label="Click to close the error modal."
          class="modal__ok-button"
        >
           ok 
        </button>
      </footer>
    </div>
  </div>
</div>
`;
